File: Grade 8/C_C++/Bootleg-McDs/Customer/customer.cpp
#include "customer.h"
using std::array, std::exception, std::map, std::numeric_limits, std::pair, std::streamsize, std::string, std::stringstream, std::vector;
using std::endl, std::find, std::getline, std::next, std::to_string, std::cin, std::cout, std::stoi;

// Constructor
Customer::Customer(string customerName) {
	// Initialize sqlite3
	stringstream ss;
	rc = sqlite3_open("customers.db", &db);
	if (rc) {
		cout << "Can't open database: " << sqlite3_errmsg(db) << "\n";
		sqlite3_close(db);
		return;
	}

	// Statement prep
	ss << "SELECT cash, prev_order, prev_price FROM customers WHERE name = '" << antiSqlInjection(customerName) << "';";
	string sql = ss.str();
	sqlite3_stmt *stmt;
	rc = sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, NULL);
	if (rc != SQLITE_OK) {
		cout << "Preparation failed: " << sqlite3_errmsg(db) << "\n";
		sqlite3_close(db);
		return;
	}

	// Load data into vector of people
	vector<CustomerData> people;
	while (sqlite3_step(stmt) == SQLITE_ROW) {
		people.push_back(
			CustomerData(
				customerName,
				sqlite3_column_double(stmt, 1),
				string(reinterpret_cast<const char *>(sqlite3_column_text(stmt, 2))),
				sqlite3_column_double(stmt, 3)));
	}

	if (people.size() == 0) {
		isNew = true;
		name = customerName;
		cash = 0;
		prevOrder.order = "1";
		prevOrder.price = 0;
		price = 0;

		// insert new customer into db
		ss.str("");
		ss << "INSERT INTO customers (name) VALUES ('" << antiSqlInjection(name) << "');";
		sql = ss.str();
		rc = sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, NULL);
		if (rc != SQLITE_OK) {
			cout << "Preparation failed: " << sqlite3_errmsg(db) << "\n";
			sqlite3_close(db);
			return;
		}
	} else {
		price = 0;
		isNew = false;
		name = customerName;
		cash = people[0].cash;
		prevOrder.order = people[0].prev_order;
		prevOrder.price = people[0].prev_price;
	}

	menu = {
		MenuItem("2 Piece chicken nuggets", 2.99),
		MenuItem("5 Piece chicken nuggets", 4.99),
		MenuItem("10 (+1) Piece chicken nuggets", 6.99),
		MenuItem("2 Piece chicken wings", 3.99),
		MenuItem("5 Piece chicken wings", 5.99),
		MenuItem("10 (+1) Piece chicken wings", 7.99),
		MenuItem("Small fries", 0.99),
		MenuItem("Medium fries", 1.99),
		MenuItem("Large fries", 2.99),
		MenuItem("Fountain soft drink", 0.99),
		MenuItem("Bottled water", 0.99),
		MenuItem("Lemonade", 1.99),
	};
}

// Ordering functions
bool Customer::order(int order) {
	// Check if order is valid
	order -= 1;
	if (order < 0 || order > 11) return false;

	cart.push_back(order);
	price += menu[order].price;
	return true;
}

bool Customer::order() {
	// convert prev_order from csv to vector<int>
	stringstream ss(prevOrder.order);
	string item;
	while (getline(ss, item, ',')) {
		cart.push_back(stoi(item));
	}

	price = prevOrder.price;
	return checkout();
}

int Customer::remove(int order) {
	// Check if order is valid
	order -= 1;
	if (order < 0 || order > 11) return 1;

	// Check if order is in cart
	vector<int>::iterator it = find(cart.begin(), cart.end(), order);
	if (it == cart.end()) return 2;

	// Remove order from cart
	cart.erase(it);

	// Update price
	price -= menu[order].price;
	return 0;
}

int Customer::checkout() {
	// Check if cart is empty
	if (cart.size() == 0) return 1;

	// Check if customer has enough cash
	if (price > cash) return 2;

	// Convert cart to csv string for db
	stringstream ss;
	for (int i = 0; i < cart.size(); i++) {
		ss << cart[i] << ",";
	}
	string order = ss.str();
	order.pop_back();

	// Update customer's cash
	cash -= price;

	// Update db
	stringstream ss2;
	ss2 << "UPDATE customers SET cash = " << cash << ", prev_order = '" << order << "', prev_price = " << price << " WHERE name = '" << name << "';";
	string sql = ss2.str();
	sqlite3_stmt *stmt;
	rc = sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, NULL);
	if (rc != SQLITE_OK) {
		cout << "Preparation failed: " << sqlite3_errmsg(db) << "\n";
		sqlite3_close(db);
		return 3;
	}

	cout << checkoutMsg() << "\n";
	return true;
}

string Customer::checkoutMsg() {
	stringstream ss;
	map<string, int> cartMap;

	// Converting cart vector to map
	for (int i = 0; i < cart.size(); i++) {
		auto orderInMenu = menu[cart[i]];
		map<string, int>::iterator it = cartMap.find(orderInMenu.name);

		// Checking if order is already present in map
		if (it == cartMap.end()) {
			cartMap.insert(pair<string, int>(orderInMenu.name, 1));
		} else {
			it->second++;
		}
	}
	// Converting map to string
	ss << "\nYou Ordered:\n";
	for (map<string, int>::iterator it = cartMap.begin(); it != cartMap.end(); it++) {
		ss << " - " << it->first;

		if (it->second > 1) {
			ss << " (x" << it->second << ")";
		}

		ss << "\n";
	}

	// Adding final messages and returning stream in form of std::string
	ss << "\n"
	   << "You spent: $" << price << "\n"
	   << "Remaining cash: $" << cash << "\n"
	   << "Thank you for your purchase! Come again soon!";

	return ss.str();
}

File: Grade 8/Python/Summative Preview/Customer.py
from cs50 import SQL

class Customer:
    # Inner Classes
    class MenuItem:
        def __init__(self, name, price):
            self.name = name
            self.price = price

        def __str__(self):
            return f"{self.name} - ${self.price}"

    # Class members
    _db = SQL("sqlite:///customers.db")
    menuItems = [
        MenuItem("2 Piece Chicken Nugget", 2.99),
        MenuItem("5 Piece Chicken Nugget", 5.99),
        MenuItem("10 (+1) Piece Chicken Nugget", 10.99),
        MenuItem("Golden Burger", 69.42),
        MenuItem("Small Fries", 1.25),
        MenuItem("Medium Fries", 2.25),
        MenuItem("Large Fries", 4.20),
        MenuItem("Waffle", 3.99),
        MenuItem("Cookie", 1.99),
        MenuItem("Ice Cream", 2.99),
        MenuItem("Fountain Drink", 1.00),
        MenuItem("Bottle of Water", 1.00),
    ]

    def __init__(self, name):
        userData = self._db.execute(
            "SELECT * FROM customers WHERE name = ?",
            name
        )

        if userData:  # Customer already exists in db
            # Set instance members
            self.name = name
            self.cash = float(userData[0]["cash"])
            self.prev_order = userData[0]["prev_order"]
            self.prev_price = float(userData[0]["prev_price"])
            self.cart = list[int]
            self.price = 0
            self.isNew = False
        else:
            self._db.execute(
                "INSERT INTO customers (name, prev_order, prev_price) VALUES (?, '0', 0)",
                name
            )  # Adding user to database

            # Set instance members
            self.name = name
            self.cash = 1000.00
            self.prev_order = "0"
            self.prev_price = 0
            self.cart = list[int]
            self.price = 0
            self.isNew = True

    def order(self, order: int) -> int:
        # Make sure order is valid
        order -= 1  # Subtract 1 to make it 0-indexed
        if (order < 0) or (order > 11):
            return -1

        # Add order to cart
        self.cart.append(order)
        self.price += self.menuItems[order].price
        return 0

    def order(self) -> int: # Function orders previous order instead of new order
        # convert order from csv to list of ints
        order = self.prev_order.split(",")
        for item in order:
            self.cart.append(int(item))
        self.price = float(self.prev_price)
        return self.checkout()

    def rm(self, order: int) -> int:
        # Make sure order is valid
        order -= 1
        if (order < 0) or (order > 11):
            return -1

        # Remove order from cart
        if order not in self.cart:
            return -2
        self.cart.remove(order)
        return 0

    def checkout(self) -> bool:
        # Make sure user has enough cash
        if self.price > self.cash:
            return False
        self.cash -= self.price

        # Convert cart to csv str
        order = ""
        for item in self.cart:
            order += str(item) + ","
        order = order[:-1] # Remove trailing comma

        # Update DB
        self._db.execute(
            "UPDATE customers SET cash = ?, prev_order = ?, prev_price = ? WHERE name = ?",
            self.cash,
            order,
            self.price,
            self.name
        )

        return True

    def getReciept(self) -> str:
        reciept = {}
        for item in self.cart:
            if item in reciept.keys():
                reciept[item] += 1
            else:
                reciept[item] = 1
        recieptStr = "\033[2;32mYou ordered:\033[0m\n\n"
        for item, quantity in reciept.items():
            recieptStr += f" - {self.menuItems[item].name} (x{quantity})\n"
        recieptStr += f"\n\033[2;32mTotal: ${self.price} \nRemaining cash: ${self.cash}\n[0m"
        recieptStr += "Have a good day!"
        return recieptStr